### ISSUE TITLE
Accessibility?

### ISSUE THREAD

**Issue by fluffynuts** [wokhan/WFN#141]
WFN is really good software, functionality-wise. It's also completely inaccessible because of the hard-coded bright colors. It would be really nice if there was a more "windows standard" theme, or, even better, some kind of theme selection, such that a user could select a higher-contrast, dark theme. I'd love to help, but unfortunately my WPF-fu is _very_ weak. I could still take a look, but I'm not promising that I can actually provide a PR.

However, I think that this would be a good thing for WFN - I've recommended it to people I work with because it's functionally _great_.

**Comment by fluffynuts** [comment]
ok, I can see I'm way out of my depth here; perhaps I'll give this a more serious go when I'm on leave :|

**Comment by fluffynuts** [comment]
I'll try have a look when I'm on leave (as of next week). A friend of mine with terrible eyesight tried it WFN and was like "nope, can't use", which I think is a pity - he's the kind of person who would get a lot of value out of it: technically-minded, works in the realm of hardware and network maintenance for a school. As stated earlier though, my WPF-fu is very weak ): I gave it a 25-min go and got... nowhere. I mean, I duplicated some brushes and so on, but couldn't manage to make them apply ): I still think WFN is aces - the kind of thing that should probably be baked into the OS. Thanks for all the time you've put into it.

**Comment by wokhan** [comment]
Hi, tell your friend I'm sorry about that and that I'll check what can be done. And a big thank you for being that supportive :-)

**Comment by wokhan** [comment]
Update: I'm working on the addition of different themes (light, dark, but also a last one more interesting for this issue, since it will rely on whatever has been set on Windows, meaning that accessibility compliant colors would be applied everywhere, along with the chosen font size). So far I'm 60% done (nothing committed yet though), I hope I'll be able to continue working on it tomorrow. Looks promising 😊

**Comment by wokhan** [comment]
This morning I added a simple way to detect the system settings (light mode / dark mode / highcontrast mode). It's not dynamic yet but still better than nothing (as changing through the WFN options would actually require being able to go in there first... not easy when the screen is not readable 😄).
I still have to tweak the colors and I'll probably need someone with eyesight issues to test this... 
@fluffynuts Would your friend be OK to give it a try once I consider it as "usable"? (should be this week).
Thanks

**Comment by fluffynuts** [comment]
I'm sure he won't mind taking a look. I'm excited for it too - I don't have real accessibility requirements, but I do find light-on-dark easier to read. Hit me up when you're ready 👍

**Comment by wokhan** [comment]
Hi, a new version is available in the "nightly" release (direct link: [x64](https://github.com/wokhan/WFN/releases/download/nightlies/wfn-x64-full-nightly-no-test-20230101-88c605.zip) / [x86](https://github.com/wokhan/WFN/releases/download/nightlies/wfn-x64-full-nightly-no-test-20230101-88c605.zip)). 

By default, it should use whatever has been set up in Windows (**dark/light mode** but also **high contrast**, which is what we're interested in here). There is also a new option to manually pick the corresponding theme (I've to improve this but wanted to release a preview before going back to work tomorrow because I won't have as much free time 😅). 

Btw I upgraded the app to .NET 7 which also paves the way for more improvements. 

I'll wait for your feedback @fluffynuts as I cannot really test the high contrast mode myself (it hurts my eyes so it's difficult for me to ensure it will help concerned users 🙄). I still have to tweak some controls but at least we now have a starting point.

Thanks!

And... Happy new year! 🥳 🍾

**Comment by fluffynuts** [comment]
this sounds absolutely epic! I'm not back at work until the 9th, so if you don't hear from me until then, that's why - i only use windows for work; however, I'm about to ping my friend to check it out too.

**Comment by fluffynuts** [comment]
just to double-check: there's no release for this right? I don't mind building to test, just want to confirm.

**Comment by wokhan** [comment]
> just to double-check: there's no release for this right? I don't mind building to test, just want to confirm.

There is, but under the "permanent" nightlies release, I embedded the links in my previous comment 😉
The assets are not ordered by date, if you want to check directly the said release, but there are 3 of them from 2023 (cannot be fresher 😁).
Thanks!

**Comment by fluffynuts** [comment]
hi @wokhan , I haven't booted up windows, but that friend of mine has given it a go and he says it doesn't start up properly, and he gets this just before it exits:

![image](https://user-images.githubusercontent.com/4669340/210259231-0ceef86c-4625-4ef7-9aaa-59cee22948ee.png)

**Comment by wokhan** [comment]
Hi @fluffynuts, thanks for the feedback! I thought I handled the first-launch case properly but I guess I didn't, so the "theme" setting is an empty string instead of null. 
On my computer I should remove the settings file everytime to simulate this but I don't... so I missed that.

I'll try to fix it tonight, but family comes first 😁

**Comment by wokhan** [comment]
Hi @fluffynuts, this issue should now be fixed. You can grab the latest nightly ([x86](https://github.com/wokhan/WFN/releases/download/nightlies/wfn-x86-full-nightly-no-test-20230103-39d0cb.zip) / [x64](https://github.com/wokhan/WFN/releases/download/nightlies/wfn-x64-full-nightly-no-test-20230103-39d0cb.zip))
Thanks

**Comment by fluffynuts** [comment]
I've finally had a chance to look at the nightly - I can see a lot of work has gone in and there are some great improvements

The things to look out for are contrast-related. Making a really good dark theme is a mission, I know (:

In particular, the more colorful areas, which work well on the light theme, can pose difficulties for sight-impaired people on the dark theme - some are difficult even for me to work with. Examples:

![image](https://user-images.githubusercontent.com/4669340/211252461-1ab3d4c2-5763-45b9-8b5d-4b00ed1cb3b5.png)

![image](https://user-images.githubusercontent.com/4669340/211252705-69187e3f-d325-4456-8891-0d81325ce7d5.png)

the things to look out for:

1. text or icons where the text color is of a close luminance to the background color - there's a formula for calculating at https://www.w3.org/TR/WCAG20-TECHS/G18.html#G18-resources
2. large areas of brighter coloring will "drown out" the rest of the UI. I don't have the condition, but a friend of mine showed me one day how, if he was using the standard Windows XP "classic" theme and a dark desktop, then the taskbar would "bleed upwards" to about 1/2 the height of the screen in his vision.

I feel like a really great theme for WFN would be one which mimics the "new" windows settings dialogs, eg:
![image](https://user-images.githubusercontent.com/4669340/211254661-f4ab89de-4b82-4b1d-a0b5-42c3e6a304c8.png)

Personally, I'd appreciate the more "minimal" interface - less color, "simpler" - but the other "sale point" I have for this is that, as I've said earlier, it feels like WFN is the kind of thing that should have been baked into Windows by Microsoft - it's a friendlier interface than the advanced firewall dialog, but also way more powerful, being able to allow down to the port/host of a connection, instead of just at the app level.

OTOH, I can see that a lot of care and effort has been put into the UI - perhaps the dark variant just needs some tweaking for contrast.

Some other issues found in testing include:
- after attempting to sort the log view:
![image](https://user-images.githubusercontent.com/4669340/211255756-cc476206-c3a4-4751-8426-affc55fa525c.png)
If I scroll a long way, i eventually find entries - it just looks like there's a lot of "empty entries" in the view?
- after switching back to the primary view and clicking a row:
![image](https://user-images.githubusercontent.com/4669340/211256065-c0a0a928-5411-49ba-b037-4c4a5fa59cda.png)
- I'm not sure if a "dark view" for the map is possible? It's easier to use in the "aerial mode", but it would be nice to have a darker map. Being able to visualise traffic is, imo, one of WFN's really really nifty features
- local network connections are reported at 0,0 on the map (if WFN finds a 0,0 co-ordinate, it could just assume that it's the same place on the map as the origin machine):
![image](https://user-images.githubusercontent.com/4669340/211256439-70e9f420-1d5f-48e5-99a3-aac01f5a76cf.png)
- the bandwidth graph is another great feature of WFN - it needs better contrast in the dark theme:
![image](https://user-images.githubusercontent.com/4669340/211256552-5e2f8bf0-d93b-4bd9-9bb5-a5f82ae7327d.png)
- I'm not sure if the bandwidth graph is monitoring the wrong interface (I have wifi and lan, but docked, I'll use lan), but even during a speedtest.net run, I couldn't get the line to budge? If there's a way to set which interface to monitor, I haven't found it

all-in-all, I still think WFN is really good, and would like to help with it in any way I can, even if that's just testing (:

**Comment by wokhan** [comment]
Hi @fluffynuts, thank you for this detailed tests report!

I'll give a look at everything you mentioned, its really helping 😊

Regarding the "accessible" part, I think I haven't been clear about how it works: in fact dark mode isn't the accessible mode, but only a new mode I added when rewriting the themes support part (because once I started that, adding color themes was easy) 😊
To see what it would look like for visually impaired users, you can set the high contrast mode on Windows itself, and WFN will use it (except if the theme has been set in the app options, in this specific case you'd have to switch back to either "system" or "automatic", the latter using whatever has been set on Windows (light / dark / high contrast).

Btw the "system" theme thumbnail is not working properly yet (I still have to find a way to fix this), but it will match the colors that have been set in Windows as well (even if high contrast mode hasn't been enabled).

Regarding the help you could provide (again), well, testing would be great as I'm really bad at it (haha) and I'd be really thankful for this 😊
Thank you for asking!!!
